Make P2 repo URL configurable

`spotless-plugin-gradle:6.17.0` broke our build because it now tries to download artifacts directly from `https://download.eclipse.org/eclipse/updates` whereas before they were resolved via the configured repositories in Gradle (i.e. an internal maven-central mirror).

Please make the P2 repo URL which is currently hardcoded in EquoBasedStepBuilder#addPlatformRepo configurable so we can swap it out with an internal mirror.

Stacktrace:

	Caused by: java.lang.RuntimeException: java.net.SocketTimeoutException: connect timed out
		at dev.equo.solstice.p2.Unchecked.wrap(Unchecked.java:25)
		at dev.equo.solstice.p2.P2Model.query(P2Model.java:133)
		at com.diffplug.spotless.extra.EquoBasedStepBuilder.get(EquoBasedStepBuilder.java:88)
		at com.diffplug.spotless.FormatterStepImpl.calculateState(FormatterStepImpl.java:58)
		at com.diffplug.spotless.LazyForwardingEquality.state(LazyForwardingEquality.java:56)
		at com.diffplug.spotless.LazyForwardingEquality.unlazy(LazyForwardingEquality.java:118)
		at com.diffplug.spotless.LazyForwardingEquality.unlazy(LazyForwardingEquality.java:124)
		at com.diffplug.gradle.spotless.SpotlessExtensionPredeclare.lambda$new$0(SpotlessExtensionPredeclare.java:39)
		at com.diffplug.gradle.spotless.SpotlessExtensionPredeclare.lambda$new$1(SpotlessExtensionPredeclare.java:33)
	[...]
	Caused by: java.net.SocketTimeoutException: connect timed out
		at okhttp3.internal.platform.Platform.connectSocket(Platform.kt:128)
		at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:295)
		at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:207)
		at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:226)
		at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106)
		at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74)
		at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255)
		at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
		at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
		at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
		at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
		at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
		at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
		at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
		at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
		at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
		at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
		at dev.equo.solstice.p2.P2Client.getBytes(P2Client.java:145)
		at dev.equo.solstice.p2.P2Client.getString(P2Client.java:116)
		at dev.equo.solstice.p2.P2Client$Folder.<init>(P2Client.java:209)
		at dev.equo.solstice.p2.P2Client.addUnits(P2Client.java:94)
		at dev.equo.solstice.p2.P2Session.populateFrom(P2Session.java:34)
		at dev.equo.solstice.p2.P2Model.queryRaw(P2Model.java:96)
		at dev.equo.solstice.p2.P2Model.query(P2Model.java:125)
		... 191 more

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -10,6 +10,8 @@ This document is intended for Spotless developers.
 We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (starting after version `1.27.0`).
 
 ## [Unreleased]
+### Added
+* Support configuration of mirrors for P2 repositories in `EquoBasedStepBuilder` ([#1629](https://github.com/diffplug/spotless/issues/1629)).
 
 ## [2.37.0] - 2023-03-13
 ### Added

--- a/lib-extra/src/main/java/com/diffplug/spotless/extra/EquoBasedStepBuilder.java
+++ b/lib-extra/src/main/java/com/diffplug/spotless/extra/EquoBasedStepBuilder.java
@@ -19,6 +19,7 @@ import java.io.File;
 import java.io.Serializable;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
 import java.util.Properties;
 
 import com.diffplug.spotless.FileSignature;
@@ -43,6 +44,7 @@ public abstract class EquoBasedStepBuilder {
 	private final ThrowingEx.Function<State, FormatterFunc> stateToFormatter;
 	private String formatterVersion;
 	private Iterable<File> settingsFiles = new ArrayList<>();
+	private Map<String, String> p2Mirrors = Map.of();
 
 	/** Initialize valid default configuration, taking latest version */
 	public EquoBasedStepBuilder(String formatterName, Provisioner mavenProvisioner, ThrowingEx.Function<State, FormatterFunc> stateToFormatter) {
@@ -57,6 +59,10 @@ public abstract class EquoBasedStepBuilder {
 
 	public void setPreferences(Iterable<File> settingsFiles) {
 		this.settingsFiles = settingsFiles;
+	}
+
+	public void setP2Mirrors(Map<String, String> p2Mirrors) {
+		this.p2Mirrors = Map.copyOf(p2Mirrors);
 	}
 
 	/** Returns the FormatterStep (whose state will be calculated lazily). */
@@ -85,7 +91,7 @@ public abstract class EquoBasedStepBuilder {
 
 	/** Creates the state of the configuration. */
 	EquoBasedStepBuilder.State get() throws Exception {
-		var query = model(formatterVersion).query(P2ClientCache.PREFER_OFFLINE, P2QueryCache.ALLOW);
+		var query = createModelWithMirrors().query(P2ClientCache.PREFER_OFFLINE, P2QueryCache.ALLOW);
 		var classpath = new ArrayList<File>();
 		var mavenDeps = new ArrayList<String>();
 		mavenDeps.add("dev.equo.ide:solstice:1.0.0");
@@ -98,6 +104,29 @@ public abstract class EquoBasedStepBuilder {
 		}
 		var jarState = JarState.preserveOrder(classpath);
 		return new State(formatterVersion, jarState, FileSignature.signAsList(settingsFiles));
+	}
+
+	private P2Model createModelWithMirrors() {
+		P2Model model = model(formatterVersion);
+		if (p2Mirrors.isEmpty()) {
+			return model;
+		}
+
+		ArrayList<String> p2Repos = new ArrayList<>(model.getP2repo());
+		p2Repos.replaceAll(url -> {
+			for (Map.Entry<String, String> mirror : p2Mirrors.entrySet()) {
+				String prefix = mirror.getKey();
+				if (url.startsWith(prefix)) {
+					return mirror.getValue() + url.substring(prefix.length());
+				}
+			}
+
+			throw new IllegalStateException("no mirror configured for P2 repository: " + url);
+		});
+
+		model.getP2repo().clear();
+		model.getP2repo().addAll(p2Repos);
+		return model;
 	}
 
 	/**

--- a/lib-extra/src/main/java/com/diffplug/spotless/extra/EquoBasedStepBuilder.java
+++ b/lib-extra/src/main/java/com/diffplug/spotless/extra/EquoBasedStepBuilder.java
@@ -94,7 +94,7 @@ public abstract class EquoBasedStepBuilder {
 		var query = createModelWithMirrors().query(P2ClientCache.PREFER_OFFLINE, P2QueryCache.ALLOW);
 		var classpath = new ArrayList<File>();
 		var mavenDeps = new ArrayList<String>();
-		mavenDeps.add("dev.equo.ide:solstice:1.0.0");
+		mavenDeps.add("dev.equo.ide:solstice:1.0.3");
 		mavenDeps.add("com.diffplug.durian:durian-swt.os:4.1.1");
 		mavenDeps.addAll(query.getJarsOnMavenCentral());
 		classpath.addAll(mavenProvisioner.provisionWithTransitives(false, mavenDeps));

--- a/plugin-gradle/CHANGES.md
+++ b/plugin-gradle/CHANGES.md
@@ -3,6 +3,16 @@
 We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (starting after version `3.27.0`).
 
 ## [Unreleased]
+### Added
+* Support configuration of mirrors for P2 repositories:
+  ```
+  spotless {
+    java {
+      eclipse().withP2Mirrors(['https://download.eclipse.org/', 'https://some.internal.mirror/eclipse'])
+    }
+  }
+  ```
+  The same configuration exists for `greclipse` and `eclipseCdt`.
 
 ## [6.17.0] - 2023-03-13
 ### Added

--- a/plugin-gradle/CHANGES.md
+++ b/plugin-gradle/CHANGES.md
@@ -4,7 +4,7 @@ We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (
 
 ## [Unreleased]
 ### Added
-* Support configuration of mirrors for P2 repositories:
+* Support configuration of mirrors for P2 repositories ([#1629](https://github.com/diffplug/spotless/issues/1629)):
   ```
   spotless {
     java {
@@ -12,6 +12,7 @@ We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (
     }
   }
   ```
+  Mirrors are selected by prefix match, for example `https://download.eclipse.org/eclipse/updates/4.26/` will be redirected to `https://some.internal.mirror/eclipse/eclipse/updates/4.26/`.
   The same configuration exists for `greclipse` and `eclipseCdt`.
 
 ## [6.17.0] - 2023-03-13

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/CppExtension.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/CppExtension.java
@@ -17,6 +17,8 @@ package com.diffplug.gradle.spotless;
 
 import static com.diffplug.gradle.spotless.PluginGradlePreconditions.requireElementsNonNull;
 
+import java.util.Map;
+
 import javax.inject.Inject;
 
 import org.gradle.api.Project;
@@ -55,6 +57,12 @@ public class CppExtension extends FormatExtension implements HasBuiltinDelimiter
 			Project project = getProject();
 			builder.setPreferences(project.files(configFiles).getFiles());
 			replaceStep(builder.build());
+		}
+
+		public EclipseConfig withP2Mirrors(Map<String, String> mirrors) {
+			builder.setP2Mirrors(mirrors);
+			replaceStep(builder.build());
+			return this;
 		}
 	}
 

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/GroovyExtension.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/GroovyExtension.java
@@ -17,6 +17,7 @@ package com.diffplug.gradle.spotless;
 
 import static com.diffplug.gradle.spotless.PluginGradlePreconditions.requireElementsNonNull;
 
+import java.util.Map;
 import java.util.Objects;
 
 import javax.inject.Inject;
@@ -98,6 +99,12 @@ public class GroovyExtension extends FormatExtension implements HasBuiltinDelimi
 			Project project = extension.getProject();
 			builder.setPreferences(project.files(configFiles).getFiles());
 			extension.replaceStep(builder.build());
+		}
+
+		public GrEclipseConfig withP2Mirrors(Map<String, String> mirrors) {
+			builder.setP2Mirrors(mirrors);
+			extension.replaceStep(builder.build());
+			return this;
 		}
 	}
 

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/JavaExtension.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/JavaExtension.java
@@ -21,6 +21,7 @@ import java.io.File;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
+import java.util.Map;
 import java.util.Objects;
 
 import javax.inject.Inject;
@@ -230,6 +231,12 @@ public class JavaExtension extends FormatExtension implements HasBuiltinDelimite
 			Project project = getProject();
 			builder.setPreferences(project.files(configFiles).getFiles());
 			replaceStep(builder.build());
+		}
+
+		public EclipseConfig withP2Mirrors(Map<String, String> mirrors) {
+			builder.setP2Mirrors(mirrors);
+			replaceStep(builder.build());
+			return this;
 		}
 
 	}

